helm/nginx-ingress-controller-app/values.yaml

```diff
@@ -26,7 +26,7 @@ configmap:
   # Increase hash table size to allow more server names for stability reasons
   server-name-hash-bucket-size: "1024"
   server-tokens: "false"
-  worker-processes: "4"
+  worker-processes: "1"
```

Comment:
Each nginx replica Pod starts number of processes
- controller
- cache
- master
- workers
and some processes get started periodically or on-demand, e.g. to load configuration changes.

Upstream default for number of worker processes is auto which means number of worker processes in each nginx replica/Pod is set to same number as detected number of CPUs. Problem is, current logic for detecting available CPUs is not cgroups aware.

So far we've been overriding worker-processes to 4. Just knowing we've been requesting 500m CPU and 600Mi memory this was obviously misconfigured. From HPA trials on adidas tenant clusters observed that each worker-process seems to keep its own copy of entire configuration derived from Ingress definitions, and on dev clusters with large amount of Ingress CRs especially during configuration validation/reload resource usage would double, and didn't fit in available requested resources.

Reducing worker-processes to just 1, and granting enough of memory and CPU made resource usage of each replica more predictable and stable with enough of capacity to store configuration even during reloads.

Currently these settings are sized up to handle largest known clusters.

With Prometheus per TC once it's availble, we'll be able to collect more metrics like nginx built in ones (connection count/rate, request rate), Ingress count/rate etc, analyze variance in resource utilization across clusters and potentially discover different usage profiles, so we'll very likely revisit these settings again in the future.

Reply:
I have two concerns here:

1. Increasing from 500m to 2G will have an impact on the allocatable memory/CPU of each worker node. We need to check if default instance type (or available instance types) have enough resources to give customer functional workers

2. If we set 2 CPUs would have more sense to set worker processes to `2` instead of `1`?
According to https://nginx.org/en/docs/ngx_core_module.html#worker_processes looks like a good initial configuration is to align CPU cores with workers processes

Reply:
1. I share your concern, actually testing this and won't merge until verifying it fits; got earlier today [ack from @puja108 on Slack](https://gigantic.slack.com/archives/CR99LSZ1N/p1581593594154700) that on-demand it's OK even to raise the minimal requirements.

2. Had same thoughts, through have to admit relatively naive experiments so far didn't see perf benefit, but did see better stability with single worker; maybe that's due to fact that it's not only worker processes that run in each nginx Pod, there's controller, master, cache, and periodically kicked off processes for validating and applying configuration updates. So 1 worker process seems now as safe start, we can always relatively easily adjust to 2 or revert to 4 and event disable HPA on-demand on case by case basis e.g. if rollout to existing production tenant clusters causes some previously not observed issues. Once metrics are available, and we collect more info/experience from the live deployments, we can set objective performance and capacity SLOs per replica and continuously ensure they are met with automated testing.

Reply:
1. What about adding an entry in changelogs, I will try to give more visibility so users are aware of implications

2. ok all good

Reply:
@pipo02mix I've update changelog, PTAL; we'll likely repeat this in some form and may even expand in legacy platform release notes.

Notice I've bumped that release with HPA enabled to be 1.6.0 instead of 1.5.0; pushed HPA back a bit to release first a regression fix for 1.4.0 + upgrade to latest nginx and that will be released as 1.5.0 instead. Regression was discovered during testing legacy WIP releases on aws that we discussed a bit earlier today in Slack #batman channel thread https://gigantic.slack.com/archives/C76JX6YLQ/p1581929367347400